File: src/device.js
'use strict';

const { execFile } = require('child_process');
const { promisify } = require('util');
const { SerialPort } = require('serialport');

const execFileAsync = promisify(execFile);

const DJI_VID = 0x2ca3;
const DJI_PID = 0x4006;
const QUECTEL_VID = 0x2c7c;
const QUECTEL_PID = 0x0125;
const SUPPORTED_USB_IDS = new Set([
  `${DJI_VID.toString(16)}:${DJI_PID.toString(16)}`,
  `${QUECTEL_VID.toString(16)}:${QUECTEL_PID.toString(16)}`,
]);
const AT_PORT_HINT = 'Quectel USB AT Port';

function normalizeVidPid(vid, pid) {
  if (vid == null || pid == null) return null;
  const v = typeof vid === 'number' ? vid : parseInt(String(vid), 16);
  const p = typeof pid === 'number' ? pid : parseInt(String(pid), 16);
  if (Number.isNaN(v) || Number.isNaN(p)) return null;
  return `${v.toString(16).toUpperCase().padStart(4, '0')}:${p
    .toString(16)
    .toUpperCase()
    .padStart(4, '0')}`;
}

function isSupportedVidPid(vid, pid) {
  const key = normalizeVidPid(vid, pid);
  if (!key) return false;
  return SUPPORTED_USB_IDS.has(key.toLowerCase());
}

async function listSerialPorts() {
  const ports = await SerialPort.list();
  return ports.map((p) => ({
    device: p.path || '',
    description: p.friendlyName || p.manufacturer || p.path || '',
    vid: p.vendorId ? parseInt(p.vendorId, 16) : null,
    pid: p.productId ? parseInt(p.productId, 16) : null,
    hwid: p.pnpId || '',
  }));
}

function findAtPort(ports) {
  const scored = [];
  for (const port of ports || []) {
    if (!port.device) continue;
    let score = 0;
    const desc = String(port.description || '').toLowerCase();
    if (desc.includes(AT_PORT_HINT.toLowerCase())) score += 50;
    if (isSupportedVidPid(port.vid, port.pid)) {
      score += 100;
      if (desc.includes('at')) score += 20;
    }
    if (score > 0) scored.push({ score, port });
  }
  if (!scored.length) return null;
  scored.sort((a, b) => b.score - a.score);
  return scored[0].port;
}

async function usbPresentWindows() {
  if (process.platform !== 'win32') {
    return { present: false, vidPid: '', raw: '' };
  }
  const ps = `
$ids = @('VID_2CA3*PID_4006*', 'VID_2C7C*PID_0125*')
$found = @()
foreach ($pat in $ids) {
  Get-PnpDevice -PresentOnly -ErrorAction SilentlyContinue |
    Where-Object { $_.InstanceId -like $pat } |
    ForEach-Object { $found += "$($_.Status)|$($_.FriendlyName)|$($_.InstanceId)" }
}
if ($found.Count -eq 0) { Write-Output 'NONE' } else { $found | ForEach-Object { Write-Output $_ } }
`.trim();

  try {
    const { stdout } = await execFileAsync(
      'powershell.exe',
      ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-Command', ps],
      { timeout: 30000, windowsHide: true, encoding: 'utf8' }
    );
    const out = (stdout || '').trim();
    if (!out || out === 'NONE') return { present: false, vidPid: '', raw: out };
    let vidPid = '';
    const upper = out.toUpperCase();
    if (upper.includes('VID_2CA3') && upper.includes('PID_4006')) vidPid = '2CA3:4006';
    else if (upper.includes('VID_2C7C') && upper.includes('PID_0125')) vidPid = '2C7C:0125';
    return { present: true, vidPid, raw: out };
  } catch (err) {
    return { present: false, vidPid: '', raw: '', error: String(err.message || err) };
  }
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function readUntilOk(port, timeoutMs = 2000) {
  const deadline = Date.now() + timeoutMs;
  let buf = Buffer.alloc(0);
  while (Date.now() < deadline) {
    const chunk = port.read();
    if (chunk && chunk.length) {
      buf = Buffer.concat([buf, chunk]);
      const text = buf.toString('utf8');
      if (/(?:^|\r?\n)(?:OK|ERROR)\r?\n?$/.test(text)) return text;
    } else {
      await sleep(40);
    }
  }
  return buf.toString('utf8');
}

async function probeAti(devicePath, baud = 115200, timeoutMs = 2000) {
  const port = new SerialPort({
    path: devicePath,
    baudRate: baud,
    autoOpen: false,
  });
  await new Promise((resolve, reject) => {
    port.open((err) => (err ? reject(err) : resolve()));
  });
  try {
    await sleep(200);
    port.flush();
    const lines = [];
    for (const cmd of ['AT', 'ATI', 'AT+GMM', 'AT+CGMR']) {
      port.write(`${cmd}\r`);
      await new Promise((resolve, reject) => {
        port.drain((err) => (err ? reject(err) : resolve()));
      });
      const text = await readUntilOk(port, timeoutMs);
      lines.push(`>>> ${cmd}\n${text.trim()}\n`);
    }
    return lines.join('\n');
  } finally {
    if (port.isOpen) await new Promise((r) => port.close(() => r()));
  }
}

function statusAsText(status) {
  const lines = [
    `USB present: ${status.usbPresent}`,
    `VID:PID: ${status.vidPid || '(none)'}`,
    `AT port: ${status.atPort || '(not found)'}`,
  ];
  if (status.ports && status.ports.length) {
    lines.push('COM / serial ports:');
    for (const p of status.ports) {
      const vp =
        p.vid != null && p.pid != null
          ? ` [${p.vid.toString(16).toUpperCase().padStart(4, '0')}:${p.pid
              .toString(16)
              .toUpperCase()
              .padStart(4, '0')}]`
          : '';
      lines.push(`  - ${p.device}: ${p.description}${vp}`);
    }
  }
  if (status.ati) {
    lines.push('--- ATI ---');
    lines.push(String(status.ati).trim());
  }
  for (const note of status.notes || []) lines.push(`note: ${note}`);
  if (status.error) lines.push(`error: ${status.error}`);
  return lines.join('\n');
}

async function detectStatus({ probe = true } = {}) {
  const status = {
    usbPresent: false,
    vidPid: '',
    atPort: null,
    ports: [],
    ati: null,
    error: null,
    notes: [],
    text: '',
  };

  if (process.platform === 'win32') {
    const usb = await usbPresentWindows();
    status.usbPresent = usb.present;
    status.vidPid = usb.vidPid || '';
    if (usb.error) status.notes.push(`PnP query: ${usb.error}`);
  } else {
    status.notes.push('PnP USB scan is Windows-only; listing serial ports only.');
  }

  const ports = await listSerialPorts();
  status.ports = ports;
  const match = findAtPort(ports);
  if (match) {
    status.atPort = match.device;
    if (!status.vidPid && match.vid != null && match.pid != null) {
      status.vidPid = normalizeVidPid(match.vid, match.pid);
      status.usbPresent = true;
    } else if (match.vid != null && match.pid != null) {
      status.usbPresent = true;
    }
  }

  if (probe && status.atPort) {
    try {
      status.ati = await probeAti(status.atPort);
    } catch (err) {
      status.error = `ATI probe failed on ${status.atPort}: ${err.message || err}`;
    }
  } else if (!status.atPort) {
    status.notes.push(
      `Install Quectel AT/DM/filter drivers, then reconnect. Looking for '${AT_PORT_HINT}'.`
    );
  }

  status.text = statusAsText(status);
  return status;
}

module.exports = {
  AT_PORT_HINT,
  DJI_VID,
  DJI_PID,
  QUECTEL_VID,
  QUECTEL_PID,
  listSerialPorts,
  findAtPort,
  detectStatus,
  probeAti,
  statusAsText,
};

File: app/device.py
"""Detect DJI Cellular Gen1 / Baiwang QDC507 (USB VID:PID 2CA3:4006) and AT port."""

from __future__ import annotations

import re
import subprocess
import sys
import time
from dataclasses import dataclass, field
from typing import Any

DJI_VID = 0x2CA3
DJI_PID = 0x4006
# Some community guides flash to Quectel identity; still usable for AT SMS.
QUECTEL_VID = 0x2C7C
QUECTEL_PID = 0x0125
SUPPORTED_USB_IDS = frozenset(
    {
        (DJI_VID, DJI_PID),
        (QUECTEL_VID, QUECTEL_PID),
    }
)
AT_PORT_HINT = "Quectel USB AT Port"


@dataclass
class PortInfo:
    device: str
    description: str = ""
    vid: int | None = None
    pid: int | None = None
    hwid: str = ""


@dataclass
class DeviceStatus:
    usb_present: bool = False
    vid_pid: str = ""
    at_port: str | None = None
    ports: list[PortInfo] = field(default_factory=list)
    ati: str | None = None
    error: str | None = None
    notes: list[str] = field(default_factory=list)

    def as_text(self) -> str:
        lines = [
            f"USB present: {self.usb_present}",
            f"VID:PID: {self.vid_pid or '(none)'}",
            f"AT port: {self.at_port or '(not found)'}",
        ]
        if self.ports:
            lines.append("COM / serial ports:")
            for p in self.ports:
                vidpid = ""
                if p.vid is not None and p.pid is not None:
                    vidpid = f" [{p.vid:04X}:{p.pid:04X}]"
                lines.append(f"  - {p.device}: {p.description}{vidpid}")
        if self.ati:
            lines.append("--- ATI ---")
            lines.append(self.ati.strip())
        for note in self.notes:
            lines.append(f"note: {note}")
        if self.error:
            lines.append(f"error: {self.error}")
        return "\n".join(lines)


def _list_serial_ports() -> list[PortInfo]:
    try:
        from serial.tools import list_ports
    except ImportError:
        return []

    result: list[PortInfo] = []
    for port in list_ports.comports():
        result.append(
            PortInfo(
                device=str(getattr(port, "device", "") or ""),
                description=str(getattr(port, "description", "") or ""),
                vid=getattr(port, "vid", None),
                pid=getattr(port, "pid", None),
                hwid=str(getattr(port, "hwid", "") or ""),
            )
        )
    return result


def find_at_port(ports: list[PortInfo] | None = None) -> PortInfo | None:
    """Prefer 'Quectel USB AT Port'; fall back to matching VID/PID serial ports."""
    ports = ports if ports is not None else _list_serial_ports()
    scored: list[tuple[int, PortInfo]] = []
    for port in ports:
        if not port.device:
            continue
        score = 0
        desc = port.description.casefold()
        if AT_PORT_HINT.casefold() in desc:
            score += 50
        if (
            isinstance(port.vid, int)
            and isinstance(port.pid, int)
            and (port.vid, port.pid) in SUPPORTED_USB_IDS
        ):
            score += 100
            if "at" in desc:
                score += 20
        if score > 0:
            scored.append((score, port))
    if not scored:
        return None
    scored.sort(key=lambda item: item[0], reverse=True)
    return scored[0][1]


def _usb_present_windows() -> tuple[bool, str]:
    """Check PnP for VID_2CA3&PID_4006 (or Quectel 2C7C:0125)."""
    if sys.platform != "win32":
        return False, ""
    ps = r"""
$ids = @('VID_2CA3*PID_4006*', 'VID_2C7C*PID_0125*')
$found = @()
foreach ($pat in $ids) {
  Get-PnpDevice -PresentOnly -ErrorAction SilentlyContinue |
    Where-Object { $_.InstanceId -like $pat } |
    ForEach-Object { $found += "$($_.Status)|$($_.FriendlyName)|$($_.InstanceId)" }
}
if ($found.Count -eq 0) { Write-Output 'NONE' } else { $found | ForEach-Object { Write-Output $_ } }
"""
    try:
        completed = subprocess.run(
            [
                "powershell",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-Command",
                ps,
            ],
            capture_output=True,
            text=True,
            timeout=30,
            check=False,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        return False, f"PnP query failed: {exc}"

    out = (completed.stdout or "").strip()
    if not out or out == "NONE":
        return False, ""
    vid_pid = ""
    if "VID_2CA3" in out.upper() and "PID_4006" in out.upper():
        vid_pid = "2CA3:4006"
    elif "VID_2C7C" in out.upper() and "PID_0125" in out.upper():
        vid_pid = "2C7C:0125"
    return True, vid_pid


def probe_ati(port: str, baud: int = 115200, timeout: float = 2.0) -> str:
    """Open AT port briefly and run AT / ATI. Raises on serial errors."""
    import serial

    lines: list[str] = []
    with serial.Serial(
        port=port,
        baudrate=baud,
        bytesize=serial.EIGHTBITS,
        parity=serial.PARITY_NONE,
        stopbits=serial.STOPBITS_ONE,
        timeout=0.2,
        write_timeout=2,
    ) as ser:
        time.sleep(0.2)
        ser.reset_input_buffer()
        for cmd in ("AT", "ATI", "AT+GMM", "AT+CGMR"):
            ser.write((cmd + "\r").encode("ascii"))
            ser.flush()
            deadline = time.monotonic() + timeout
            buf = bytearray()
            while time.monotonic() < deadline:
                waiting = ser.in_waiting
                chunk = ser.read(waiting if waiting else 1)
                if chunk:
                    buf.extend(chunk)
                    text = buf.decode("utf-8", errors="replace")
                    if re.search(r"(?:^|\r?\n)(?:OK|ERROR)\r?\n?$", text):
                        break
                else:
                    time.sleep(0.05)
            text = buf.decode("utf-8", errors="replace")
            lines.append(f">>> {cmd}\n{text.strip()}\n")
    return "\n".join(lines)


def detect_status(*, probe: bool = True) -> DeviceStatus:
    status = DeviceStatus()
    if sys.platform == "win32":
        present, vid_pid = _usb_present_windows()
        status.usb_present = present
        status.vid_pid = vid_pid
    else:
        status.notes.append("PnP USB scan is Windows-only; listing serial ports only.")

    ports = _list_serial_ports()
    status.ports = ports
    match = find_at_port(ports)
    if match:
        status.at_port = match.device
        if not status.vid_pid and match.vid is not None and match.pid is not None:
            status.vid_pid = f"{match.vid:04X}:{match.pid:04X}"
            status.usb_present = True
        elif match.vid is not None and match.pid is not None:
            status.usb_present = True

    if probe and status.at_port:
        try:
            status.ati = probe_ati(status.at_port)
        except Exception as exc:  # noqa: BLE001 — surface to UI
            status.error = f"ATI probe failed on {status.at_port}: {exc}"
    elif not status.at_port:
        status.notes.append(
            f"Install Quectel AT/DM/filter drivers, then reconnect. Looking for '{AT_PORT_HINT}'."
        )
    return status


def list_ports_summary() -> list[dict[str, Any]]:
    return [
        {
            "device": p.device,
            "description": p.description,
            "vid": p.vid,
            "pid": p.pid,
        }
        for p in _list_serial_ports()
    ]
